### PR TITLE
Paypal express checkout

### PR DESCRIPTION
**Pull request by ilanarielbaron**
Paypal express checkout module for hyva checkout.

Im using the paypal express magento graphQL approach. https://devdocs.magento.com/guides/v2.4/graphql/payment-methods/paypal-express-checkout.html

Two things to check:
1- The confirmation page is hardcoded in the paypal express hook as 'checkout/onepage/success' 
2- When i place the order is calling to setPaymentMethodOnCart before it goes to paypal and is not well, im using this mutation in the paypal customized way so we need to avoid that.

### CODE REVIEW

File: src/reactapp/src/api/paypalExpress/createCustomerToken/createCustomerToken.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { CREATE_PAYPAL_EXPRESS_TOKEN } from './mutation';
+import sendRequest from '../../sendRequest';
+import LocalStorage from '../../../utils/localStorage';
+import modifier from './modifier';
+
+export default async function createCustomerToken({ returnUrl, cancelUrl }) {
```
Comment (progammer-rkt) [inline src/reactapp/src/api/paypalExpress/createCustomerToken/createCustomerToken.js:6]: We are not doing anything with the parameters inside the function. So destructuring them is unnecessary here. a variable like `actionUrls` or `urlInfo` would suffice. Then below `const variables = { ...actionUrls, cartId }`.

File: src/reactapp/src/api/paypalExpress/createCustomerToken/mutation.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+export const CREATE_PAYPAL_EXPRESS_TOKEN = `
+mutation createPaypalExpressToken(
+    $cartId: String!
+    $returnUrl: String!
+    $cancelUrl: String!
+  ) {
+    createPaypalExpressToken(
+      input: {
+        cart_id: $cartId
+        code: "paypal_express"
```
Comment (progammer-rkt) [inline src/reactapp/src/api/paypalExpress/createCustomerToken/mutation.js:10]: I know the payment method code is not going to change. Still, we should avoid hard coding `paypal_express` here. We should pass it in `variable` here. This way we can make sure our mutation works even if the implementation changes the value in future.

File: src/reactapp/src/api/paypalExpress/setPaymentMethod/mutation.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    mutation setPaymentMethodOnCart(
+    $payerId: String!
+    $token: String!
+    $cartId: String!
+  ) {
+    setPaymentMethodOnCart(
+      input: {
+        cart_id: $cartId,
+        payment_method: {
+          code: "paypal_express"
```
Comment (progammer-rkt) [inline src/reactapp/src/api/paypalExpress/setPaymentMethod/mutation.js:11]: Should not hard code as stated above

File: src/reactapp/src/api/paypalExpress/setPaymentMethod/setPaymentMethod.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { SET_PAYMENT_METHOD_ON_CART_PAYPAL_EXPRESS } from './mutation';
+import sendRequest from '../../sendRequest';
+import LocalStorage from '../../../utils/localStorage';
+import modifier from './modifier';
+
+export default async function setPaymentMethodPaypalExpress({
```
Comment (progammer-rkt) [inline src/reactapp/src/api/paypalExpress/setPaymentMethod/setPaymentMethod.js:6]: No need for destructuring.

File: src/reactapp/src/components/paypalExpress/components/PaypalExpress.jsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import usePaypalExpress from '../hooks/usePaypalExpress';
+import useCheckoutFormContext from '../../../hook/useCheckoutFormContext';
+
+function PaypalExpress({ method, selected, actions }) {
+  const isSelected = method.code === selected.code;
+  const { authorizeUser } = usePaypalExpress();
+  const { registerPaymentAction } = useCheckoutFormContext();
+
+  useEffect(() => {
+    registerPaymentAction('paypal_express', authorizeUser);
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/components/PaypalExpress.jsx:13]: No need to hard code here. pass `method.code` here

File: src/reactapp/src/components/paypalExpress/components/PaypalExpress.jsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  }, [authorizeUser, registerPaymentAction]);
+
+  if (!isSelected) {
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/components/PaypalExpress.jsx:16]: The JSX inside this if condition and the JSX outside if condition are same. So basically the `if` condition is no longer necessary here.

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+import setPaymentMethodPaypalExpress from '../../../api/paypalExpress/setPaymentMethod';
+import placeOrder from '../../../api/cart/placeOrder';
+import useShippingMethodCartContext from '../../shippingMethod/hooks/useShippingMethodCartContext';
+import LocalStorage from '../../../utils/localStorage';
+import { config } from '../../../config';
+import usePaymentMethodCartContext from '../../paymentMethod/hooks/usePaymentMethodCartContext';
+import useBillingAddressCartContext from '../../billingAddress/hooks/useBillingAddressCartContext';
+
+export default function usePaypalExpress() {
+  const [, appActions] = useContext(AppContext);
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:15]: I like to keep state declaration first, then custom hooks, then following other variable declarations inside a function. This way we are consistent with other components and that will increase the readability (at least for me)

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  const { selectedPaymentMethod, methodList } = usePaymentMethodCartContext();
+  const cartId = LocalStorage.getCartId();
+
+  /*
+   Check if is possible to proceed on placing the order.
+   */
+  useEffect(() => {
+    if (
+      query &&
+      selectedMethod?.methodCode &&
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:30]: `selectedMethod` is confusing here. we are inside a payment context. so `selectedMethod` sounds like a `selectedPaymentMethod`. But we have already a variable with that name. So we need to rename `selectedMethod` => `selectedShippingMethod`. Also, we are referring the code like `selectedMethod?.methodCode` which is OK. But it will be more good if we are consistent with other parts of the code. so `const selectShippingMethodCode = _get(selectedMethod, 'methodCode')` would be an ideal choice. Same goes with `selectedPaymentMethod?.code`. Declaring these two variables outside `useEffect` and then use the variable inside `useEffect`

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+      (selectedPaymentMethod?.code === '' ||
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:31]: may be we can combine this condition like `['', 'paypal_express'].includes(selectedPaymentMethodCode)`. Remember again `paypal_express` hardcode here. we need to pass `method.code` to the hook and use it.

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        selectedPaymentMethod?.code === 'paypal_express')
+    )
+      setProcessPaymentEnable(true);
+  }, [query, setProcessPaymentEnable, selectedMethod, selectedPaymentMethod]);
+
+  useEffect(() => {
+    const processPaypalExpress = async () => {
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:38]: It is better we define `processPaypalExpress` outside `useEffect`. We need to wrap the function with `useCallback` in order to avoid unwanted re-execution of `useEffect`.

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+  return {
+    authorizeUser,
+  };
+}
+
+/*
+ Utility to get the token and the payer id from the URL
+ */
+const getTokenPayerId = query => {
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:129]: No need for this. Use native `URLSearchParams` to get the parameters. see this link: https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams

With this, we can completely get rid off this helper function and retrieve token and payer id within the hook itself.

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+import { config } from '../../../config';
+import usePaymentMethodCartContext from '../../paymentMethod/hooks/usePaymentMethodCartContext';
+import useBillingAddressCartContext from '../../billingAddress/hooks/useBillingAddressCartContext';
+
+export default function usePaypalExpress() {
+  const [, appActions] = useContext(AppContext);
+  const { setErrorMessage, setPageLoader } = appActions;
+  const { selectedMethod } = useShippingMethodCartContext();
+  const { cartBillingAddress } = useBillingAddressCartContext();
+  const query = window.location.href.split('?')[1];
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:19]: `window.location.search` will be better here to retrieve the search parameters

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+  }, [query, setProcessPaymentEnable, selectedMethod, selectedPaymentMethod]);
+
+  useEffect(() => {
+    const processPaypalExpress = async () => {
+      setProcessPaymentEnable(false);
+      /*
+       Get paypal express required data from response query
+       */
+      const { token, PayerID: payerId } = getTokenPayerId(query);
+      const paymentMethod = _objToArray(methodList).find(
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:44]: Totally unnecessary. we have `method` available in `PaypalExpress` component. So we can pass that to this hook. Also, `paymentMethod`  checking here is totally unnecessary. because paypal express option will be rendered in the frontend only if `methodList` holds paypal_express.

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+import useShippingMethodCartContext from '../../shippingMethod/hooks/useShippingMethodCartContext';
+import LocalStorage from '../../../utils/localStorage';
+import { config } from '../../../config';
+import usePaymentMethodCartContext from '../../paymentMethod/hooks/usePaymentMethodCartContext';
+import useBillingAddressCartContext from '../../billingAddress/hooks/useBillingAddressCartContext';
+
+export default function usePaypalExpress() {
+  const [, appActions] = useContext(AppContext);
+  const { setErrorMessage, setPageLoader } = appActions;
+  const { selectedMethod } = useShippingMethodCartContext();
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:17]: `useShippingMethodCartContext`, `useBillingAddressCartContext` & `usePaymentMethodCartContext` - these should not be used here. we created those custom hooks to use inside `shippingMethod`, `billingAddress`, `paymentMethod` respectively. So the thumb rule here is, we need to create custom hooks for each section so that it serves them locally. This way, the code is separate and we can easily remove the code chunks (if in case). It also helps to avoid the growing of the custom hook. 

So here we need to define custom hooks in `src/reactapp/src/components/paypalExpress/hooks/` directory. The name should be `usePaypalExpressCartContext` and then you need to prepare data in that custom hook from the `CartContext`. 

So basically `useShippingMethodCartContext`, `useBillingAddressCartContext` & `usePaymentMethodCartContext` custom hooks are collecting data from the `CartContext`. So our `usePaypalExpressCartContext` should essentially replace three of them

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+        method => method.code === 'paypal_express'
+      );
+
+      if (
+        !token ||
+        !payerId ||
+        !cartId ||
+        !selectedMethod ||
+        !paymentMethod ||
+        !cartBillingAddress
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:54]: Instead of passing the whole `cartBillingAddress`, here we need a `boolean` check. this we can do inside `usePaypalExpressCartContext`. the variable can be `hasCartBillingAddress =  billingAddress.firstName && billingAddress.country`. Passing the whole billingAddress is a risk because that state is prone to change from other actions. but making it boolean will definitely avoid re-rendering of `useCallback` that wraps this function

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+      /*
+       Set payment method on cart and place the order
+       */
+      try {
+        setPageLoader(true);
+        await setPaymentMethodPaypalExpress({ payerId, token });
+        const response = await placeOrder();
+        if (response && response.order_number) {
+          LocalStorage.clearCheckoutStorage();
+          window.location.replace(
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:68]: I think your concern is here. Yes, for this I will come up with a solution. Here we need to consider many things. When we use the checkout as an app, we need to totally avoid redirection as the success page is not available at that time. we need to perform redirection only when this works inside a site. I will look into this 👍

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    setErrorMessage,
+    setPageLoader,
+    cartBillingAddress,
+  ]);
+
+  /*
+   Get the customer token from the BE and redirect to paypal.
+   */
+  const authorizeUser = useCallback(async () => {
+    if (!selectedMethod?.carrierCode || !cartBillingAddress?.firstname) {
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:99]: Try to use `_get` function as i mentioned above please

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    setPageLoader(true);
+    const response = await createCustomerToken({
+      returnUrl: `checkout/index/index`,
+      cancelUrl: `checkout/index/index`,
+    });
+
+    if (response) {
+      const paypalExpressUrl =
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:111]: use `_get` please

File: src/reactapp/src/context/Cart/order/actions.js [hunk 19]
```diff
@@ -25,7 +25,7 @@ async function verifyPaymentMethod(dispatch, values, selectedPaymentMethod) {
     const paymentMethod = _get(values, PAYMENT_METHOD_FORM);
 
     if (!paymentMethod.code) {
-      throw new Error('Pyament method not available');
+      throw new Error('Payment method not available');
```
Comment (progammer-rkt) [inline src/reactapp/src/context/Cart/order/actions.js:28]: Good catch. Add translation here please. Also apply translation to all messages in this PR. Dont forget to add this in the file `src/reactapp/public/i18n/en_US.json`

**Review by progammer-rkt — CHANGES_REQUESTED** [review]

File: src/reactapp/src/api/paypalExpress/createCustomerToken/createCustomerToken.js (hunk 1 above)
Comment (ilanbaron) [reply src/reactapp/src/api/paypalExpress/createCustomerToken/createCustomerToken.js:6]: Got it, changed!

File: src/reactapp/src/api/paypalExpress/createCustomerToken/mutation.js (hunk 2 above)
Comment (ilanbaron) [reply src/reactapp/src/api/paypalExpress/createCustomerToken/mutation.js:10]: Done

File: src/reactapp/src/api/paypalExpress/setPaymentMethod/mutation.js (hunk 3 above)
Comment (ilanbaron) [reply src/reactapp/src/api/paypalExpress/setPaymentMethod/mutation.js:11]: Done

File: src/reactapp/src/api/paypalExpress/setPaymentMethod/setPaymentMethod.js (hunk 4 above)
Comment (ilanbaron) [reply src/reactapp/src/api/paypalExpress/setPaymentMethod/setPaymentMethod.js:6]: Done

File: src/reactapp/src/components/paypalExpress/components/PaypalExpress.jsx (hunk 5 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/components/PaypalExpress.jsx:13]: Done

File: src/reactapp/src/components/paypalExpress/components/PaypalExpress.jsx (hunk 6 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/components/PaypalExpress.jsx:16]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 7 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:15]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 8 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:30]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 9 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:31]: Great, done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 10 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:38]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 11 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:129]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 12 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:19]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 13 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:44]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 14 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:17]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 15 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:54]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 16 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:68]: Sure, lets make that change after we have our own success page

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 17 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:99]: Done

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 18 above)
Comment (ilanbaron) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:111]: Done

File: src/reactapp/src/context/Cart/order/actions.js (hunk 19 above)
Comment (ilanbaron) [reply src/reactapp/src/context/Cart/order/actions.js:28]: DOne

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js (hunk 11 above)
Comment (progammer-rkt) [reply src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:146]: Can we move `getTokenPlayerId` above the hook function? Since this helper function uses inside the hook, somone expect the definition of this helper function above the hook.

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpressCartContext.js [hunk 20]
```diff
@@ -0,0 +1,36 @@
+import { useContext, useMemo } from 'react';
+import _get from 'lodash.get';
+import CartContext from '../../../context/Cart/CartContext';
+import AppContext from '../../../context/App/AppContext';
+
+export default function usePaypalExpressCartContext() {
+  const [cartData] = useContext(CartContext);
+  const [, appActions] = useContext(AppContext);
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpressCartContext.js:8]: We should only deal with `CartContext` here. We need another hook like `usePaypalExpressAppContext`  to deal with `AppContext`

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpressCartContext.js [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+  return useMemo(() => {
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpressCartContext.js:10]: I understood why you wrapped this with `useMemo`. I was doing the same previously. But this has no benefit here. Because the dependencies here are prone to change more often which makes this callback called always all time. 

Better remove `useMemo` here

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpressCartContext.js [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    const cart = _get(cartData, 'cart');
+
+    const cartBillingAddress = _get(cart, `billing_address`);
+    const hasCartBillingAddress = !!cartBillingAddress?.firstname;
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpressCartContext.js:14]: We can add some more required fields here.

File: src/reactapp/src/context/Cart/order/actions.js (hunk 19 above)
Comment (progammer-rkt) [reply src/reactapp/src/context/Cart/order/actions.js]: Can we add this to string `src/reactapp/public/i18n/en_US.json` please. There are many other translation strings in this commit that we need to add into `en_US.json`

File: src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+    selectedShippingMethod,
+    selectedPaymentMethod,
+    setPageLoader,
+    setErrorMessage,
+    cartId,
+  } = usePaypalExpressCartContext();
+  const query = window.location.search;
+  const selectedShippingMethodCode = _get(selectedShippingMethod, 'methodCode');
+  const selectedPaymentMethodCode = _get(selectedPaymentMethod, 'code');
+  const { code } = paymentMethod;
```
Comment (progammer-rkt) [inline src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js:24]: Again `code` is not descriptive. Also, we don't want to pass the `payment method entirely. Seems this hook only uses the code. So pass only the code to the hook. maybe we can use a variable `paypalExpressCode` which clearly explains what it is.

**Review by progammer-rkt — CHANGES_REQUESTED** [review]
We are almost there!!! You nailed it.

Please build the app once you did the changes. for that, you need to run `npm run build`.

File: src/reactapp/public/i18n/en_US.json [hunk 24]
```diff
@@ -61,5 +61,8 @@
   "Total":"Total",
   "Update":"Update",
   "{} zipcode: {}":"{} zipcode: {}",
+  "Payment method not available": "Payment method not available",
+  "Shipping method not available": "Shipping method not available",
+  "Please complete all the required data": "Please complete all the required data"
```
Comment (progammer-rkt) [inline src/reactapp/public/i18n/en_US.json:66]: There is a mistake in the json. You forgot comma here :)

**Comment by progammer-rkt** [comment]
We decided to move this work to a separate repository: https://github.com/hyva-themes/magento2-hyva-checkout-paypal-express

Hence closing this pull request.

**Comment by TheRealHunzik** [comment]
@progammer-rkt  di i need to install another module to use Paypal express with hyva checkout?
